DEVELOPERS.md

```diff
@@ -0,0 +1,421 @@
+
+It is assumed the modifications and compilation will be done on the [NCI HPC system `gadi`](https://opus.nci.org.au/display/Help/Gadi+User+Guide).
+
+Much of this adapted from the [spack Developer Workflows Tutorial](https://spack-tutorial.readthedocs.io/en/latest/tutorial_developer_workflows.html).
+
+## Initial set-up 
+
+The steps in this section need only be done once, as long as the versions of the three repositories are sufficient.
+
+It is necessary to make a local copy of `spack`, the configuration suitable for `gadi` (`spack-config`) and a package repository (`spack-packages`) which defines a number of the packages used by ACCESS-OM2.
+
+Adapted from [instructions on the forum](https://forum.access-hive.org.au/t/how-to-build-access-om2-on-gadi/1545)
+
+```bash
+git clone -c feature.manyFiles=true https://github.com/ACCESS-NRI/spack.git --branch releases/v0.21 --single-branch --depth=1
+git clone https://github.com/ACCESS-NRI/spack-packages.git --branch main
+git clone https://github.com/ACCESS-NRI/spack-config.git --branch main
+```
+
+Note: `spack` is cloned from the ACCESS-NRI fork as on occasion there are some fixes that are back-ported by ACCESS-NRI.
+
+Add all the `spack-config` settings to your local `spack` instance:
+```bash
+ln -s -r -v spack-config/v0.21/gadi/* spack/etc/spack/
+```
+
+## Enable spack
+
+For every new login or new shell it is necessary to add `spack` to your shell, as well as some other settings. 
+Do note the space between the `.` and the path, as we are not treating `spack-enable.bash` as an executable:
+```bash
+. spack-config/spack-enable.bash
+```
+
+## Development environment
+
+Spack [has environments](https://spack.readthedocs.io/en/latest/environments.html), which is similar to many other related implementations such as conda environments and python virtual environments.
+
+Environments create an isolated operating environment within which `spack` can only see and access packages that are added to the environment. In this repository the model to be built is defined using an environment in the `spack.yaml` file.
+
+Spack generally automatically downloads package sources to a cache, compiles and installs them.  Spack also supports model development using a modified environment where the components that are being actively developed are compiled in a local source directory, and the rest built in the normal spack manner.
+
+### Source code
+
+There are two options for local source directory location: 
+
+1. Use a path to an existing source code repository
+2. Let spack clone the code for you and place it in the environment directory
+
+If option 1 is preferred the source code for the component(s) to be modified has to be available on the filesystem, e.g. using `git clone`, before the next steps. For more details, see the [Notes](#notes) section below.
```

Comment:
```suggestion
If option 1 is preferred the source code for the component(s) to be modified has to be available on the filesystem, e.g. from using `git clone`, before the next steps. For more details, see the [Notes](#notes) section below.
```

If you talk of "preferred", should this document expand a bit on which option is best for what case? I must say it seems pretty obvious to me but just saying. 

A quick example of where it isn't clear: if we are planning to do some developments but haven't started, is it better to choose option 1 directly so everything is set to test our local dev before making them public? Or would you go with option 2 first and then modify to option 1 if/when necessary? This might be obvious after reading the whole doc, but it isn't just when reaching this point.

Reply:
I don't think it is clear which is the best approach, but agreed making some points about that might be useful. I'm not sure I'm in the best position to know all those, but broadly it depends if you want to keep all your source code changes in a single location (repo) on disk, or have different local repo copies for each environment.

Before something like `spack environments` came along it was a no-brainer that it would be a single repo. 

Now we have the possibility of separated build environments it might make more sense to have a repo per environment and keep code isolated to the build environment.

The two approaches are entirely equivalent, but the single repo would require some extra steps to ensure the code repo is pointing to the correct branch when an environment is activated.

The repo-per-environment approach doesn't require any effort to keep the build and code in sync, but it does involve copies of code, and some overhead keeping repos in sync if that is necessary.

So I don't think it is cut n' dried, and might depend a lot on how long lived feature branches are, and how much interaction is required with other feature branches.

Reply:
I have added this note:

> The choice of where the source code for an environment should reside depends on use case and personal prefrence. If an existing source code repository location is used in more than one environment it requires the code repository be kept in sync with the purpose of the development environment, e.g. two separate feature branches utilising two different development environments would require the developer to check out the correct branch when switching between development environments. It would also preclude building both environments simultaneously.